language/move-prover/tests/sources/functional/verify_vector.move

```diff
@@ -291,6 +291,23 @@ module 0x42::VerifyVector {
         ensures old(v[i]) == result;
     }
 
+    // Add an element to the vector at index.
+    fun verify_model_insert<Element>(v: &mut vector<Element>, e: Element, i: u64) {
+        vector::insert(v, e, i); // inlining the built-in Boogie procedure
+    }
+    spec verify_model_insert {
+        aborts_if i > len(v);
+        ensures len(v) == len(old(v)) + 1;
+        ensures v[i] == e;
+        ensures old(v[0..i]) == v[0..i];
+        // the following ensures is so complicated because we cannot use len(old(v)) in a nested
+        // context and have to rely on len(v) to give us length of the old vector
+//        ensures len(old(v)) == 0
```

Comment:
I am not sure why this `ensures` clause does not work. I have been staring at this code for too long to be able to easily figure out if the problem is with the clause or my implementation, and I was hoping for the second pair of eyes @junkil-park and @meng-xu-cs ...

Reply:
This ensures can be simplified into
```
        ensures v[i+1..len(v)] == old(v)[i..len(old(v))];
```

The overall spec might be formatted into:
```
    spec verify_model_insert {
        aborts_if i > len(v);
        ensures len(v) == len(old(v)) + 1;
        ensures v[0..i] == old(v)[0..i];
        ensures v[i] == e;
        ensures v[i+1..len(v)] == old(v)[i..len(old(v))];
    }
```

Reply:
> This ensures can be simplified into
> 
> ```
>         ensures v[i+1..len(v)] == old(v)[i..len(old(v))];
> ```
Indeed, thank you! I missed the fact that you can do `old(v)[x..y]` in addition to `old(v[x..y])`... 

It also seems like the problem was with the `ensures` clause as with the suggested update the test is working!

Reply:
@meng-xu-cs , @junkil-park - there is an error in a file that does not even use `vector::insert`. Do you have any suggestions on how to deal with it?

Reply:
What error is it in what file?